controllers/openstackdataplanenode_controller.go

```diff
@@ -107,36 +107,39 @@ func (r *OpenStackDataPlaneNodeReconciler) Provision(ctx context.Context, instan
 	return nil
 }
 
-// Inventory struct
-// TODO: make use of the struct below
-//
-//nolint:unused
+// Host represents ansible host
+type Host struct {
+	Vars map[string]string `yaml:"vars,omitempty"`
+}
+
+// Group represents ansible group
+type Group struct {
+	Vars  map[string]string `yaml:"vars,omitempty"`
+	Hosts map[string]*Host  `yaml:"hosts,omitempty"`
+}
+
+// Inventory contains parsed inventory representation
 type Inventory struct {
-	all struct {
-		hosts struct {
-			host struct {
-				hostVar struct {
-					hostVarValue string
-				}
-			}
-		}
-	}
+	Groups map[string]*Group
 }
 
 // GenerateInventory yields a parsed Inventory
 func (r *OpenStackDataPlaneNodeReconciler) GenerateInventory(ctx context.Context, instance *dataplanev1beta1.OpenStackDataPlaneNode) error {
 	var err error
 
-	inventory := make(map[string]map[string]map[string]map[string]string)
-	all := make(map[string]map[string]map[string]string)
-	host := make(map[string]map[string]string)
-	hostVars := make(map[string]string)
-	hostVars["ansible_host"] = instance.Spec.Node.HostName
-	hostVars["ansible_user"] = instance.Spec.Node.AnsibleUser
-	hostVars["ansible_port"] = strconv.Itoa(instance.Spec.Node.AnsiblePort)
-	host[instance.Name] = hostVars
-	all["hosts"] = host
-	inventory["all"] = all
+	host := Host{}
+	host.Vars = make(map[string]string)
+	host.Vars["ansible_host"] = instance.Spec.Node.HostName
+	host.Vars["ansible_user"] = instance.Spec.Node.AnsibleUser
+	host.Vars["ansible_port"] = strconv.Itoa(instance.Spec.Node.AnsiblePort)
+
+	group := Group{}
+	group.Hosts = make(map[string]*Host)
+	group.Hosts[instance.Name] = &host
+
+	inventory := Inventory{}
+	inventory.Groups = make(map[string]*Group)
+	inventory.Groups["all"] = &group
```

Comment:
This addresses my initial concerns about building a string vs a struct. Thanks, I think that's much better. 

I'm curious what @SeanMooney and @olliewalsh think.

Reply:
in terms of a minimal implemnation this is certenly much better then what was there before.

its a lot more readable 

i was originally using map[string]interface{}

so that if you wanted to set something to an int bool or null you could

currently, you don't have the ability to do that as all your types are converted to strings so i would personally prefer to do that over map[string]string for the vars

both i prefer this vs the current state prior to this pr.

Reply:
the main delta vs what ollie proposed in https://go.dev/play/p/92S1JDkHGDW
is that host and groups don't know there name and groups cant be nested.

ollies example is more complete in that respect but i don't really like that it need the fixnames function.

over all i think i prefer https://go.dev/play/p/92S1JDkHGDW but I'm not against this either.

Reply:
You mean I can't get away with a teeny-tiny little hack like the fixnames function :smiley:

FWIW I've a slightly better version that uses a custom unmarshaller to call fixnames - https://go.dev/play/p/AuJqU08RPmS

+1 for using `map[string]interface{}`. Can get around it with Itoa for the ansible_port but will be necessary I expect if more complex vars need to be set such as NetworkConfig

Reply:
I've followed @olliewalsh suggestions here: https://github.com/openstack-k8s-operators/dataplane-operator/pull/17